pkg/services/ngalert/schedule/schedule.go

```diff
@@ -549,9 +347,21 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key models.AlertRul
 
 		processedStates := sch.stateManager.ProcessEvalResults(ctx, r, results)
 		sch.saveAlertStates(ctx, processedStates)
-		alerts := FromAlertStateToPostableAlerts(processedStates, sch.stateManager, sch.appURL)
-
-		notify(alerts, logger)
+		toNotify := make([]*state.State, 0, len(processedStates))
+		sent := sch.clock.Now()
+		for _, alertState := range toNotify {
+			if !alertState.NeedsSending(sch.stateManager.ResendDelay) {
```

Comment:
I think this should be in the notifier rather than the scheduler?

Reply:
I think the notifier should not decide whether a state needs to be sent or not. I think ideally, that should be decided by the state manger. However, as long as this is refactoring, I try to limit scope of the changes.

Reply:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent, but the state manager is responsible to deciding if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.

Reply:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent and the state manager should decide if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.

Reply:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent and the state manager should decide if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.

Reply:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent and the state manager should decide if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.

Reply:
I suppose we have different opinions on where the separation should be.

To me a notifier should decide if an alert needs to be sent and the state manager should decide if a state should be turned into an alert. These sound similar, but I think there is an important distinction to be made here.